stdlib/src/tempfile/tempfile.mojo

```diff
@@ -164,3 +164,84 @@ fn mkdtemp(
         except:
             continue
     raise Error("Failed to create temporary file")
+
+
+# TODO use shutil.rmtree (or equivalent) when it exists
+fn _rmtree(path: String, ignore_errors: Bool = False) raises:
+    """Removes the specified directory and all its contents.
+    If the path is a symbolic link, an error is raised.
+    If ignore_errors is True, errors resulting from failed removals will be ignored.
+    Absolute and relative paths are allowed, relative paths are resolved from cwd.
+    Args:
+      path: The path to the directory.
+      ignore_errors: Whether to ignore errors.
+    """
+    if os.path.islink(path):
+        raise Error("`path`can not be a symbolic link")
+
+    for file_or_dir in os.listdir(path):
+        if os.path.isfile(path + "/" + file_or_dir[]):
+            try:
+                os.remove(path + "/" + file_or_dir[])
+            except e:
+                if not ignore_errors:
+                    raise Error(e)
+            continue
+        if os.path.isdir(path + "/" + file_or_dir[]):
+            try:
+                _rmtree(path + "/" + file_or_dir[], ignore_errors)
+            except e:
+                if ignore_errors:
+                    continue
+                raise Error(e)
+    try:
+        os.rmdir(path)
+    except e:
+        if not ignore_errors:
+            raise Error(e)
+
+
+struct TemporaryDirectory:
+    """A temporary directory."""
+
+    var name: String
+    """The name of the temporary directory."""
+    var _ignore_cleanup_errors: Bool
+    """Whether to ignore cleanup errors."""
+
+    fn __init__(
+        inout self,
+        suffix: String = "",
+        prefix: String = "tmp",
```

Comment:
In Python these can also be `None`. Can we also use `Optional[String] = None` for these?

Reply:
`Optional[String]` does not play along nicely with `StringLiteral`, that's why I used default values for these, it should be possible to allow to use `None` but I think we will need 4 overloads to cover all the cases:
```
fn __init__(inout self, suffix: String = "", prefix: String = "tmp", ...)
fn __init__(inout self, suffix: Optional[String] = None, prefix: String = "tmp", ...)
fn __init__(inout self, suffix: String = "", prefix: Optional[String] = None, ...)
fn __init__(inout self, suffix: Optional[String] = None, prefix: Optional[String] = None, ...)
```

Reply:
Yeah, this is a known issue, but workaround with the overloads can be simpler:
```mojo
fn __init__(inout self, suffix: String, prefix: String = "tmp", ...)
fn __init__(inout self, suffix: NoneType = None, prefix: String = "tmp", ...)
fn __init__(inout self, suffix: String = "", prefix: NoneType = None, ...)
fn __init__(inout self, suffix: NoneType = None, prefix: NoneType = None, ...)
```

Reply:
Could you please address this in a followup?